libcxx/include/typeinfo

```diff
@@ -275,13 +275,15 @@ struct __type_info_implementations {
           __impl;
 };
 
-#    if defined(__arm64__) && __has_cpp_attribute(clang::ptrauth_vtable_pointer)
-#      if __has_feature(ptrauth_type_info_discriminated_vtable_pointer)
+#    if __has_cpp_attribute(_Clang::__ptrauth_vtable_pointer__) && __has_feature(__ptrauth_calls__)
+#      if __has_feature(__ptrauth_vtable_address_discrimination__) ||                                                  \
+          __has_feature(__ptrauth_vtable_type_discrimination__)
```

Comment:
A different signing schema is used for type info vtable pointers compared to other vtable pointers: see https://github.com/ahmedbougacha/llvm-project/blob/eng/arm64e-upstream-llvmorg/clang/include/clang/Basic/PointerAuthOptions.h#L246 (also see a comment in code for context). The reason is that in `ItaniumRTTIBuilder::BuildVTablePointer`, we only have mangled names of classes derived from `std::type_info`, and do not have full definitions with info about base class. So, we can't use type-dependent extra discrimination, since for vtable pointers, it relies on ability to look up base class definition. As for address discrimination for `type_info`-related vtable pointers, it could probably be supported on clang side, but it currently isn't.

Given that, in libcxx, previously, we unconditionally disabled type-based and address discrimination for type info vtable pointers. Here is the condition we used in downstream:

```
#if (defined(__arm64__) || defined(__aarch64__)) && \
    __has_feature(ptrauth_calls) && \
    __has_cpp_attribute(clang::ptrauth_vtable_pointer)
#define _LIBCPP_TYPE_INFO_VTABLE_POINTER_AUTH \
   [[clang::ptrauth_vtable_pointer(process_independent, no_address_discrimination, no_extra_discrimination)]]
#else
#define _LIBCPP_TYPE_INFO_VTABLE_POINTER_AUTH
#endif
```

I see that now type-based and address discrimination are enabled in libcxx under certain conditions: see https://github.com/ahmedbougacha/llvm-project/blob/eng/arm64e-upstream-llvmorg/libcxx/include/typeinfo#L278

```
#    if defined(__arm64__) && __has_cpp_attribute(clang::ptrauth_vtable_pointer)
#      if __has_feature(ptrauth_type_info_discriminated_vtable_pointer)
#        define _LIBCPP_TYPE_INFO_VTABLE_POINTER_AUTH                                                                  \
          [[clang::ptrauth_vtable_pointer(process_independent, address_discrimination, type_discrimination)]]
#      else
#        define _LIBCPP_TYPE_INFO_VTABLE_POINTER_AUTH                                                                  \
          [[clang::ptrauth_vtable_pointer(process_independent, no_address_discrimination, no_extra_discrimination)]]
#      endif
#    else
#      define _LIBCPP_TYPE_INFO_VTABLE_POINTER_AUTH
#    endif
```

@ahmedbougacha I was unable to find any information about `ptrauth_type_info_discriminated_vtable_pointer` preprocessor feature in your branch (except its usage in include/typeinfo in libcxx). I also do not see any changes related to additional discrimination in initialization of `CXXTypeInfoVTablePointer` field of `class PointerAuthOptions`. It looks like there are no related changes in `ItaniumRTTIBuilder::BuildVTablePointer` as well. Please let me know if non-zero discrimination is actually supported on clang side for type info vtable pointers?

@ahmedbougacha @asl @ldionne I think there are 3 questions we need to answer to construct a condition everyone likes.

1. Do we want to check target architecture? If so, both `__arm64__` and `__aarch64__` should be checked. I'd prefer to have this check - even though it's probably redundant since other ptrauth-related checks are true only on aarch64, IMHO for those who don't know all the context it might be useful to see the architecture name and know that "OK, it's aarch64-specific stuff".

2. Do we want to check `ptrauth_calls` preprocessor feature enabled? I'd prefer to have this check - it's probably also redundant since `[[clang::ptrauth_vtable_pointer]]` attribute should have no effect w/o `ptrauth_calls` preprocessor feature, but, again, more verbosity might be useful for those who reads the code.

3. Do we want to allow type-based and/or address discrimination for type info vtable pointers? If so, what's the condition for that? Check against `__has_feature(ptrauth_type_info_discriminated_vtable_pointer)` like in @ahmedbougacha 's branch, or smth else?

Reply:
@ldionne @ahmedbougacha @asl
**Update.** My preference is the following:
1. Do not check arm64/aarch64. If in future we have ptrauth support on non-arm64, for example, via soft pointer auth, the conditions will remain the same. Since we don't have different logic for ptrauth+non-arm64, no need for additional check.
2. Do not check against `ptrauth_calls`. The `[[clang::ptrauth_vtable_pointer]]` attribute has no effect w/o `ptrauth_calls` enabled.
3. If we want to introduce `__has_feature(ptrauth_type_info_discriminated_vtable_pointer)`, @ahmedbougacha please update your branch to include changes allowing such discrimination - now there is no support on clang side.

Reply:
Yes, we shouldn't do this change here, it does the opposite of what we want:  the ptrauth (current arm64e, or elf pauthabi) ABI does sign vtable pointers with type/address discrimination.  But it specifically doesn't sign the type info vtable pointer, which is why we need this special annotation, to override the default vtable ptr discrimination (that's exposed with the `ptrauth_vtable_address_discrimination`/`ptrauth_vtable_type_discrimination` features) to match the type info vtable codegen.

Since then we have indeed added support for signing these, and that's indeed denoted by `ptrauth_type_info_discriminated_vtable_pointer`.  So I think the version that's in tree today is doing the right thing, and honoring clang's directive given its understanding of the ABI.  (I don't mind and I'm not worried about removing the arm64 check FWIW).

We'll open a clang PR to add `ptrauth_type_info_discriminated_vtable_pointer`.  But looking at it now I think we'll want to rename it first, so let's settle that this week before it's too late (cc @ojhunt)

@kovdan01 to answer your specific questions:  in general I do think these should check something else than ptrauth features, because the specific decisions are made by a specific platform ABI, and isn't set in stone for all ptrauth.  Concretely, for the soft ptrauth example, a real soft ptrauth implementation could very well force these to be always discriminated, and use a different key (soft ptrauth not being constrained by HW keys, it uses a different key for every single ABI usage;  it could use one for "type info vtable ptrs").
In practice there likely won't ever be a serious and meaningfully different ptrauth ABI, so this doesn't seem like a real problem either way.

Reply:
@ahmedbougacha As far as I understood, you are OK with current implementation present in this PR. It enables type+address discrimination for type info vtable pointers when we have `__has_feature(__ptrauth_vtable_address_discrimination__) || __has_feature(__ptrauth_vtable_type_discrimination__)`. Is this really the thing we want to? I'm a bit concerned with this since I've never seen support for discriminating type info vtable pointers (although I understand that you have that already), and the condition present in the PR will enable such discrimination while mainline clang does not support that yet.

Previously, the condition was `__has_feature(ptrauth_type_info_discriminated_vtable_pointer)`, and I think it's better since it does not affect type info vtable pointers discrimination unless it's really implemented on clang side.

Reply:
> Previously, the condition was __has_feature(ptrauth_type_info_discriminated_vtable_pointer), and I think it's better since it does not affect type info vtable pointers discrimination unless it's really implemented on clang side.

Yes, to be clear I agree that's what's best